Conch/lib/Conch/Validation.pm

```diff
@@ -0,0 +1,625 @@
+validation results are built as C<HASHREF>s. The C<CODEREF> will be passed a
+list of attributes and values (Attributes are 'message', 'name', 'category',
+'status', and 'hint') whenever a validation result is created.
+
+=back
+
+=cut
+
+sub new ( $class, %attrs ) {
+
+	bless {
+		_device                    => $attrs{device},
+		_device_location           => $attrs{device_location},
+		_hardware_product          => $attrs{hardware_product},
+		_device_settings           => $attrs{device_settings} || {},
+		_validation_result_builder => $attrs{result_builder} || sub { return {@_} },
+		validation_results         => []
+	}, $class;
+
+}
+
+=head2 run
+
+Run the Validation with the specified input data.
+
+	$validation->run($validation_data);
+
+=cut
+
+sub run ( $self, $data ) {
+
+	try { $self->run_unsafe($data) }
+	catch {
+		my $frame = $_->frames->[0];
+
+		# Provide module name and line number in error message, not filepath
+		my $error_loc =
+			'Exception raised in \'' . $frame->[0] . '\' at line ' . $frame->[2];
+		my $validation_error = $self->{_validation_result_builder}->(
+			message  => $_->message,
+			name     => $self->name,
+			status   => STATUS_ERROR,
+			hint     => $error_loc,
+			category => $self->category,
+		);
+		push $self->validation_results->@*, $validation_error;
+	};
+}
+
+=head2 run_unsafe
+
+Run the Validation with the specified input data. Re-throws any exception raised
+during execution as L<Mojo::Exception>
+
+	eval {
+		$validation->run_unsafe($validation_data)
+	};
+	say $@->frames if $@;
+
+=cut
+
+sub run_unsafe ( $self, $data ) {
+	local $SIG{__DIE__} = sub {
+		my $err = shift;
+		if ( $err->isa('Mojo::Exception') ) {
+			CORE::die $err;
+		}
+		else {
+			# remove the 'at $filename line $line_number' from the exception
+			# message. We might not want to reveal Conch's path
+			$err =~ s/ at .+$//;
+			$self->die( $err, 2 );
+		}
+	};
+
+	$self->check_against_schema($data);
+	$self->validate($data);
+}
+
+=head2 check_against_schema
+
+Check the Validation input data against the JSON schema, if specified.
+
+=cut
+
+sub check_against_schema ( $self, $data ) {
+	return unless $self->schema;
+	my $schema_validator = JSON::Validator->new;
+
+	# Shallow copy the schema and pull out the 'required' list
+	my $schema   = { $self->schema->%* };
+	my $required = delete $schema->{required};
+	my $full_schema =
+		{ type => 'object', properties => $schema, required => $required };
+
+	$schema_validator->schema($full_schema);
+	my @errors = $schema_validator->validate($data);
+	$self->die( "@errors", 2 ) if @errors;
+}
+
+=head2 validate
+
+Contains the validation logic for validations.
+
+This method must be re-defined in sub-classes of C<Conch::Validation> or it will
+raise an exception.
+
+	package MyValidation;
+	use Mojo::Base 'Conch::Validation';
+
+	sub validate {
+		my ($self, $data) = @_;
+		$self->register_result({ expected => 1, got => $data->{pass} });
+	}
+
+=cut
+
+sub validate ( $self, $data ) {
```

Comment:
I originally intended to make it abundantly clear to someone writing a validation for the first time. It blows up and they're told what they need to fix it.

I also use it as a hook to satisfy pod.t. With the `Pod::Coverage::CountParents` option that I set, Test::Pod::Coverage is satisfied if a method overridden in the child is present in the parent. If `validate` isn't defined here, it will be required. I don't want to require pod docs beyond copyright in the validations. The description fulfills describing what the validation is and does.